File: infrastructure/lib/observability-stack.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { Stack, StackProps } from 'aws-cdk-lib'
import { Construct } from 'constructs';
import { Color, Dashboard, GraphWidget, Metric } from 'aws-cdk-lib/aws-cloudwatch';

export class ObservabilityStack extends Stack {

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const disconnectionsMetric = new Metric({
      namespace: 'websocket-chat',
      metricName: 'closedConnection',
      statistic: 'sum'
    });

    const newcConnectionsMetric = new Metric({
      namespace: 'websocket-chat',
      metricName: 'newConnection',
      statistic: 'sum'
    });

    const messagesDeliveredMetric = new Metric({
      namespace: 'websocket-chat',
      metricName: 'messageDelivered',
      statistic: 'sum'
    });


    var closedConnectionsWidget = new GraphWidget({
      title: "Closed Connections",
      width: 12,
      left: [disconnectionsMetric.with({
        color: Color.RED
      })]
    });

    var newConnectionsWidget = new GraphWidget({
      title: "New Connections",
      width: 12,
      left: [newcConnectionsMetric.with({
        color: Color.GREEN
      })]
    });

    var messagesDeliveredWidgets = new GraphWidget({
      title: "Messages Delivered",
      width: 24,
      left: [messagesDeliveredMetric.with({
        color: Color.GREEN
      })]
    });

    const dashboard = new Dashboard(this, "Serverless Websocket Chat Dashboard", {
      widgets: [
        [
          newConnectionsWidget,
          closedConnectionsWidget
        ],
        [
          messagesDeliveredWidgets,
        ]
      ]
    });
  }
};

File: infrastructure-net7/src/Infrastructure/Stacks/ObservabilityStack.cs
using System.Collections.Generic;
using Amazon.CDK;
using Amazon.CDK.AWS.CloudWatch;
using Constructs;

namespace Infrastructure.Stacks
{
    public class ObservabilityStack : Stack
    {

        internal ObservabilityStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
        {
            #region Metrics
            var disconnectionsMetric = new Metric(new MetricProps()
            {
                Namespace = "websocket-chat",
                DimensionsMap = new Dictionary<string, string>() {{"Service", "service_undefined"}},
                MetricName = "closedConnection",
                Statistic = "sum"
            });
            
            var newConnectionsMetric = new Metric(new MetricProps()
            {
                Namespace = "websocket-chat",
                DimensionsMap = new Dictionary<string, string>() {{"Service", "service_undefined"}},
                MetricName = "newConnection",
                Statistic = "sum"
            });
            
            var messagesDeliveredMetric = new Metric(new MetricProps()
            {
                Namespace = "websocket-chat",
                DimensionsMap = new Dictionary<string, string>() {{"Service", "service_undefined"}},
                MetricName = "messageDelivered",
                Statistic = "sum"
            });
            #endregion
            
            #region Widgets
            var closedConnectionsWidget = new GraphWidget(new GraphWidgetProps()
            {
                Title = "Closed Connections",
                Width = 12,
                Left = new[]{ disconnectionsMetric.With(new MetricOptions()
                {
                    Color = Color.RED
                })}
            });
            
            var newConnectionsWidget = new GraphWidget(new GraphWidgetProps()
            {
                Title = "New Connections",
                Width = 12,
                Left = new[]{ newConnectionsMetric.With(new MetricOptions()
                {
                    Color = Color.GREEN
                })}
            });
            
            var messagesDeliveredWidgets = new GraphWidget(new GraphWidgetProps()
            {
                Title = "Messages Delivered",
                Width = 24,
                Left = new[]{ messagesDeliveredMetric.With(new MetricOptions()
                {
                    Color = Color.GREEN
                })}
            });
            #endregion

            var dashboard = new Dashboard(this, "Serverless Websocket Chat Dashboard", new DashboardProps()
            {
                Widgets = new []
                {
                    new []
                    {
                        newConnectionsWidget, 
                        closedConnectionsWidget
                    },
                    new []
                    {
                        messagesDeliveredWidgets
                    }
                }
            });
        }
    }
}
